src/voter.rs

```diff
@@ -869,10 +966,64 @@ impl<H, N, E: Environment<H, N>> Future for Voter<H, N, E> where
 		};
 
 		self.past_rounds.push(background);
-		self.committer.push(old_number, old_round.clone());
+		self.committer.push(old_round_number, old_round);
 
-		// round has been updated. so we need to re-poll.
-		self.poll()
+		Ok(())
+	}
+
+	fn completed_prospective_round(&mut self, mut prospective_round: VotingRound<H, N, E>)
+		-> Result<(), E::Error>
+	{
+		self.env.completed(prospective_round.votes.number(), prospective_round.votes.state())?;
+
+		self.best_round = VotingRound::new(
+			prospective_round.votes.number() + 1,
+			// FIXME: this may be behind, the `prospective_round` won't issue
+			// finalization notifications (since it doesn't vote)
```

Comment:
i'll ask alistair what to do here; it's possible we could take `prospective_round.votes.state().finalized` if it's higher than `last_finalized_in_rounds`

Reply:
Or we could at least use the finalized commit target that triggered the prospective round.

Reply:
(it's also not the worst thing if this is behind but it also could be a bit of a performance hit if we have to construct large DAGs)

Reply:
let's leave as is for now. we should just optimize the `vote_graph` and `ancestry` if this ends up being too slow in practice

Reply:
> Or we could at least use the finalized commit target that triggered the prospective round.

this sounds right